### PR TITLE
Script for splitting large zip files

### PR DESCRIPTION
**Pull request by cezara98t**

### CODE REVIEW

File: datamanager/splitLargeZip/splitZip.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import argparse
+import csv
+import os
+import tempfile
+import zipfile
+from dataclasses import dataclass
+from io import TextIOWrapper
+from pathlib import Path
+
+ONEGB = 1000000000
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
ONE_GB = 2 ** 30
```

File: datamanager/splitLargeZip/splitZip.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    path_to_schema = ""
+    path_to_split = ""
+    start_size = 0  # size of schema and split file
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--path", help="the path of the zip")
+    args = parser.parse_args()
+    processZip(args.path)
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: try/except and proper error message if it fails

File: datamanager/splitLargeZip/splitZip.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+"""
+This script is used to split zip files (DM exports) that are above 1GB in size.
+
+Usage:
+    splitZip.py --path C:\\path\\to\\zip
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: let's use snake_case for both variables and file names as this is Python standard:
https://www.python.org/dev/peps/pep-0008/
very short version: https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html

File: datamanager/splitLargeZip/splitZip.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import zipfile
+from dataclasses import dataclass
+from io import TextIOWrapper
+from pathlib import Path
+
+ONEGB = 1000000000
+
+
+@dataclass
+class configs:
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
class Configs:
```

File: datamanager/splitLargeZip/splitZip.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    path_to_schema = ""
+    path_to_split = ""
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
    schema_path: Path  = Path()
    split_path: Path = Path()
```

File: datamanager/splitLargeZip/splitZip.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    start_size = 0  # size of schema and split file
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
    start_size: int = 0  # size of schema and split file
```

File: datamanager/splitLargeZip/splitZip.py (hunk 1 above)
Comment (bogdancondurache) [reply datamanager/splitLargeZip/splitZip.py]: or maybe even better: `ONE_GB = 2 ** 30 - 1`

File: datamanager/splitLargeZip/splitZip.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser()
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
    parser = argparse.ArgumentParser(description="Split a dataset into multiple archives.")
```

File: datamanager/splitLargeZip/splitZip.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    parser.add_argument("--path", help="the path of the zip")
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
    parser.add_argument("--path", help="the path to the zip file")
```

File: datamanager/splitLargeZip/splitZip.py (hunk 3 above)
Comment (bogdancondurache) [reply datamanager/splitLargeZip/splitZip.py]:
```suggestion
    split_zip.py --path C:\\path\\to\\import\\archive.zip
```

File: datamanager/splitLargeZip/splitZip.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    args = parser.parse_args()
+    processZip(args.path)
+
+
+def processZip(path):
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
def process_zip(path: Path):
```

File: datamanager/splitLargeZip/splitZip.py (hunk 2 above)
Comment (bogdancondurache) [reply datamanager/splitLargeZip/splitZip.py]:
```suggestion
    processZip(Path(args.path))
```

Comment (bogdancondurache) [reply datamanager/splitLargeZip/splitZip.py]:
```suggestion
    process_zip(args.path)
```

File: datamanager/splitLargeZip/splitZip.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    with tempfile.TemporaryDirectory() as temp_zip_contents:
+        with zipfile.ZipFile(path, "r") as zip_ref:
+            zip_ref.extractall(temp_zip_contents)
+            dictionary_paths = {
+                path.name: path for path in list(Path(temp_zip_contents + "/" + path.rstrip(".zip")).iterdir())
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
                path.name: path for path in Path(temp_zip_contents) / path.stem).iterdir()
```
you don't need list conversion as `for ... in` knows to user generators and iterators. Also if `path` would be of type Path you can use `.stem` instead os rstrip which is more hackish. Path joining can be done even as in the suggestion with `/` or with `.joinpath()`.
Please check https://docs.python.org/3/library/pathlib.html - it also has alternatives from using strings to using Path objects. Also this for iterators and generators: https://anandology.com/python-practice-book/iterators.html

File: datamanager/splitLargeZip/splitZip.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            }
+
+            configs.path_to_schema = dictionary_paths["schema.json"]
+            configs.start_size += os.stat(configs.path_to_schema).st_size
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
            configs.start_size += configs.path_to_schema.stat().st_size
```
if this is Path object

File: datamanager/splitLargeZip/splitZip.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+            configs.path_to_split = dictionary_paths["split.csv"]
+            configs.start_size += os.stat(configs.path_to_split).st_size
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: same as above

File: datamanager/splitLargeZip/splitZip.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+            splitFiles(dictionary_paths["images"], dictionary_paths["latest"], temp_zip_contents)
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
            split_files(dictionary_paths["images"], dictionary_paths["latest"], temp_zip_contents)
```

File: datamanager/splitLargeZip/splitZip.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def splitFiles(images_path, latest_path, temp_zip):
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
def split_files(images_path: Path, latest_path: Path, temp_zip: Type?):
```

File: datamanager/splitLargeZip/splitZip.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    image_list = list()
+    temp_size = configs.start_size
+    arhive_number = 1
+    dict_image_paths = {path.name: path for path in list(Path(images_path).iterdir())}
+    dict_latest_paths = {path.name.strip(".json"): path for path in list(Path(latest_path).iterdir())}
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: `.stem`

File: datamanager/splitLargeZip/splitZip.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+    for path_name in dict_image_paths:
+        size = os.stat(dict_image_paths[path_name]).st_size + os.stat(dict_latest_paths[path_name]).st_size
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: `.stat().st_size`

File: datamanager/splitLargeZip/splitZip.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        else:
+            writeOneZip(temp_zip, image_list, arhive_number)
+            arhive_number += 1
+            temp_size = configs.start_size + size
+            image_list = list()
+            image_list.append(path_name)
+    writeOneZip(temp_zip, image_list, arhive_number)  # for the last zip
+
+
+def writeOneZip(temp_zip, image_list, archive_number):
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
def write_one_zip(temp_zip, image_list, archive_number):
```
and types please

File: datamanager/splitLargeZip/splitZip.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    with zipfile.ZipFile("splitArchive" + str(archive_number) + ".zip", "w") as zip_handle:
+        for root, _, files in os.walk(temp_zip):
+            for file in files:
+                for image_path in image_list:
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: could you use `zip()` for these?

File: datamanager/splitLargeZip/splitZip.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+                    if image_path in file:
+                        zip_handle.write(
+                            os.path.join(root, file),
+                            os.path.relpath(os.path.join(root, file), temp_zip),
+                        )
+        zip_handle.write(
+            os.path.join(root, configs.path_to_schema),
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]: let's use pathlib instead :)

File: datamanager/splitLargeZip/splitZip.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+    with open(configs.path_to_split, "rb") as f:
+        reader = csv.reader(TextIOWrapper(f, "utf-8"), delimiter="\t")
+        lines.append(next(reader))
+        for row in reader:
+            if row[0] in image_list:
+                lines.append(row)
+
+    new_split_name = str(configs.path_to_split) + "_"
+    with open(new_split_name, "w", newline="") as writeFile:
```
Comment (bogdancondurache) [inline datamanager/splitLargeZip/splitZip.py]:
```suggestion
    with open(new_split_name, "w", newline="") as write_file:
```

**Review by bogdancondurache — CHANGES_REQUESTED** [review]

File: datamanager/splitLargeZip/splitZip.py (hunk 15 above)
Comment (misupantea-uipath) [reply datamanager/splitLargeZip/splitZip.py]: `archive_number`

**Review by misupantea-uipath — APPROVED** [review]

File: datamanager/splitLargeZip/split_zip.py [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+            Configs.start_size += Configs.split_path.stat().st_size
+
+            split_files(top_level_paths["images"], top_level_paths["latest"], temp_zip_contents)
+
+
+def split_files(images_path: Path, latest_path: Path, folder_contents_path: str):
+    image_names = list()
+    current_size = Configs.start_size
+    archive_number = 1
+    dict_image_paths = {path.name: path for path in list(Path(images_path).iterdir())}
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: `image_paths`

File: datamanager/splitLargeZip/split_zip.py [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    dict_latest_paths = {path.stem: path for path in list(Path(latest_path).iterdir())}
+
+    for image_name in dict_image_paths:
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: enumerate key-value pairs?

File: datamanager/splitLargeZip/split_zip.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+        else:
+            write_one_zip(folder_contents_path, image_names, archive_number)
+            archive_number += 1
+            current_size = Configs.start_size + size
+            image_names = list()
+            image_names.append(image_name)
+    write_one_zip(folder_contents_path, image_names, archive_number)  # for the last zip
+
+
+def write_one_zip(folder_contents_path: str, image_names: list, archive_number: int):
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: `write_one_zip` -> `create_archive`

File: datamanager/splitLargeZip/split_zip.py [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+            Configs.split_path = top_level_paths["split.csv"]
+            Configs.start_size += Configs.split_path.stat().st_size
+
+            split_files(top_level_paths["images"], top_level_paths["latest"], temp_zip_contents)
+
+
+def split_files(images_path: Path, latest_path: Path, folder_contents_path: str):
+    image_names = list()
+    current_size = Configs.start_size
+    archive_number = 1
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: `archive_count`

File: datamanager/splitLargeZip/split_zip.py (hunk 23 above)
Comment (misupantea-uipath) [reply datamanager/splitLargeZip/split_zip.py]: `archive_number` -> `suffix`

File: datamanager/splitLargeZip/split_zip.py [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+            write_one_zip(folder_contents_path, image_names, archive_number)
+            archive_number += 1
+            current_size = Configs.start_size + size
+            image_names = list()
+            image_names.append(image_name)
+    write_one_zip(folder_contents_path, image_names, archive_number)  # for the last zip
+
+
+def write_one_zip(folder_contents_path: str, image_names: list, archive_number: int):
+    with zipfile.ZipFile("splitArchive" + str(archive_number) + ".zip", "w") as zip_handle:
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: `"splitArchive"` -> maybe use the original archive name

Use `_` for suffixing

File: datamanager/splitLargeZip/split_zip.py [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+        for root, _, files in os.walk(folder_contents_path):
+            for file in files:
+                for image_name in image_names:
+                    if image_name in file:
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: Add a comment here to point out that both the image and related metadata file will be picked up for writing due to the `in` check

File: datamanager/splitLargeZip/split_zip.py [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+    lines = list()
+
+    with open(Configs.split_path, "rb") as f:
+        reader = csv.reader(TextIOWrapper(f, "utf-8"), delimiter="\t")
+        lines.append(next(reader))
+        for row in reader:
+            if row[0] in images:
+                lines.append(row)
+
+    temp_split_path = str(Configs.split_path) + "_"
```
Comment (misupantea-uipath) [inline datamanager/splitLargeZip/split_zip.py]: Use the `suffix` param from `create_archive` to generate the `temp_split_path`, that way we can debug splits easier

**Review by misupantea-uipath — APPROVED** [review]

**Review by bogdancondurache — APPROVED** [review]

**Pull request merged**